torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -597,29 +597,38 @@ void CudaCodeGen::call(const std::vector<CallArg>& args) {
   USE_TRIGGER(cuda_codegen_executed);
 }
 
+void CudaSetContext(CUcontext pctx) {
+  if (!pctx) {
+    std::unique_lock<std::mutex> cudaFreeMutexLock(
+        *(c10::cuda::CUDACachingAllocator::getFreeMutex()));
+    cudaFree(0);
+  }
+}
+
 void CudaCodeGen::CompileToNVRTC(
     const std::string& code,
     const std::string& func_name) {
   // Initializes driver's API context (if necessary)
   CUdevice device = 0;
   CUcontext pctx = 0;
+
   AT_CUDA_DRIVER_CHECK(nvrtc().cuCtxGetCurrent(&pctx));
-  if (!pctx) {
-    std::unique_lock<std::mutex> cudaFreeMutexLock(
-        *(c10::cuda::CUDACachingAllocator::getFreeMutex()));
-    cudaFree(0);
-  }
+  CudaSetContext(pctx);
 
   // Note: hacked at::DeviceGuard since at::DeviceGuard was failing to work
   // properly in some scenarios
   const auto prior_device = at::cuda::current_device();
-  at::cuda::set_device(device);
+  at::cuda::set_device(this->device().index());
+  // cudaSetDevice does not have to really change the underlying device if it
+  // doesn't have to, so calling cudaFree to force that change
+  CudaSetContext(pctx);
```

Comment:
I still don't think this is necessary here (above, maybe, it looks like it's forcing a driver context initialization).  Maybe @zheng-xq can weigh in.

Reply:
Since this changes from the old device to the new device, uninitialized context could happen to either of them. Also I am starting to wondering why we need to initialize the old device, and not only the new device?

The strange thing here is "pctx" is from the old device. Protonu, is that really what you want?

Reply:
Thanks, XQ. Made the change.

Reply:
Are you not supposed to "cuCtxGetCurrent(&pctx))" after "at::cuda::set_device"?

Reply:
Yes, I missed that, updated

Reply:
At this point, I don't think CudaSetContext is necessary now. Alternatively, fold "CUcontext pctx = 0" and "cuCtxGetCurrent" into a "CudaForceContextInit".